code-review/README.md

```diff
@@ -0,0 +1,119 @@
+Code Review
+===========
+
+This is forked/stolen from ^W^W inspired by
+[Thoughtbot's code review guide](https://raw.githubusercontent.com/thoughtbot/guides/master/code-review/README.md)
+
+A guide for reviewing code and having your code reviewed.
+
+Why code reviews?
+-----------------
+
+Code is reviewed before being merged into `master`. The goal of review is to ensure that changes are delivering business value to our customers - not to perfect our code.
+
+When reviewing, consider whether a change:
+
+* Delivers business value. Understand why it is being made and why it is required.
+* Does what it says it does. Try to break it.
+* Can be maintained by the most junior member of our team.
+* Meets our standards and is well written.
+
+Don't let minor technical issues block shipping.
+
+A reviewer may give +1 on a review. Small fixes need +1, complex fixes and large features +2. A branch should have at least +1 before you hit the green button.
+
+Code Review Mindset
+-------------------
+
+It's important to separate our self-worth from the code we write so that we are always open to feedback that will help us improve.
```

Comment:
This appears to be a paraphrasing of a main point from Weinberg's _Egoless Programming_ (1971). Jeff "CodingHorror" Atwood extracts [10 "commandments" here](http://blog.codinghorror.com/the-ten-commandments-of-egoless-programming/). I think it's an extreme oversimplification to extract just this one principle and omit others. In particular, the point summarized in the tenth commandment is:

> _Critique code instead of people – be kind to the coder, not to the code. As much as possible, make all of your comments positive and oriented to improving the code. Relate comments to local standards, program specs, increased performance, etc._

On the other hand, Atwood also mentions how _egoless programming_ has been considered [a fallacy or myth](http://blog.codinghorror.com/you-gotta-own-it/). The tl;dr here is that we *want* people to be personally invested in (to _own_) their work: we want their passion, commitment:

> _if you do want great software, you have to let the developers own what they're building. The developers are inevitably the ones who have the most control over the success or failure of the project. Creating an environment where your developers have no emotional attachment to the project they're working on is a recipe for mediocre software-- and job disillusionment._

So the paradox of peer review is that we expect developers to own/invest/care deeply for the work they do, right up until the point that their work is reviewed, when they should magically adopt a zen-like disinterest in having it reviewed. I submit that this is unreasonable, *unless* we're very careful about how reviews are done, and what reviewing *is*.

Karl Wiegers wrote an excellent book in 2001 called [Peer Reviews in Software: A Practical Guide] (http://www.amazon.com/dp/0201734850/?tag=stackoverfl08-20) which looked at the psychological aspects of peer review. Chapter two [is available online](http://www.processimpact.com/reviews_book/chapter_2.pdf), and here's a relevant quote (my emphasis):

> _The dynamics between the work product’s author and its reviewers are critical. The
author must trust and respect the reviewers enough to be receptive to their comments.
Similarly, the reviewers must show respect for the author’s talent and hard work.
Reviewers should **thoughtfully select the words they use** to raise an issue, focusing on
what they observed about the product. Saying, “I didn’t see where these variables were
initialized” is likely to elicit a constructive response, whereas “You didn’t initialize these
variables” might get the author’s hackles up. The small shift in wording from the
accusatory “you” to the **less confrontational** “I” lets the reviewer deliver even critical
feedback effectively. Reviewers and authors must continue to work together outside the
reviews, so they all need to maintain a level of professionalism and mutual respect to
avoid strained relationships._

The reason that I quote these at this point is because where I have in the past seen code reviews go quite spectacularly wrong, it's been because of (1) a lack of attention to these points of personal interaction, and (2) reviews focusing on subjective attributes of code style or system design. The two factors multiply together and have bad effects:
* the review drifts off from looking for bugs in the code into abstract discussion of points of style, practice or formatting (which are generally not decideable in the scope of a code review, if ever)
* because the discussion ceases to be based on objective truths (_this code has no test_, _this code is brittle_, _this exception isn't caught_) it becomes an argument, and humans (especially male humans) feel that arguments need to be *won*. Objective issues with code are much, much easier to accept.

Above in this document, the statement is made that _The goal of review is to ensure that changes are delivering business value to our customers - not to perfect our code_.  For me, this (excellent principle) means that reviews are intended to spot defects and ensure that code meets the business-value requirements: and in particular, that *subjective opinions about how code should be written* are not in agreement with that goal.

tl;dr - I don't think we pay nearly enough attention to the personal interaction aspects of code review, and subjective opinions exacerbate this problem because they change discussion into argument.